lib/agents.js

```diff
@@ -1,199 +1,203 @@
 'use strict'
 
-const http = require('http')
-const https = require('https')
 const net = require('net')
 const tls = require('tls')
 const { once } = require('events')
-const { createTimeout, abortRace, urlify, appendPort, cacheAgent } = require('./util')
+const timers = require('timers/promises')
+const { urlify, appendPort } = require('./util')
 const { normalizeOptions, cacheOptions } = require('./options')
-const { getProxy, getProxyType, proxyCache } = require('./proxy.js')
+const { getProxy, getProxyAgent, proxyCache } = require('./proxy.js')
 const Errors = require('./errors.js')
+const { Agent: AgentBase } = require('agent-base')
 
-const createAgent = (base, name) => {
-  const SECURE = base === https
-  const SOCKET_TYPE = SECURE ? tls : net
+module.exports = class Agent extends AgentBase {
+  #options
+  #timeouts
+  #proxy
 
-  const agent = class extends base.Agent {
-    #options
-    #timeouts
-    #proxy
-    #socket
+  constructor (options) {
+    const { timeouts, proxy, noProxy, ...normalizedOptions } = normalizeOptions(options)
 
-    constructor (_options) {
-      const { timeouts, proxy, noProxy, ...options } = normalizeOptions(_options)
+    super(normalizedOptions)
 
-      super(options)
+    this.#options = normalizedOptions
+    this.#timeouts = timeouts
 
-      this.#options = options
-      this.#timeouts = timeouts
-      this.#proxy = proxy ? { proxies: getProxyType(proxy), proxy: urlify(proxy), noProxy } : null
+    if (proxy) {
+      this.#proxy = {
+        proxy: urlify(proxy),
+        noProxy,
+        Agent: getProxyAgent(proxy),
+      }
```

Comment:
This needs some better naming. We have class `Agent` with nested properties `Agent.#proxy.proxy`, `Agent.#proxy.agent`, and `Agent.#proxy.noProxy`. It's not at all clear what these fields should mean.

Reply:
And similarly we have the `Agent.proxy` getter and the `Agent.getProxy()` function.

Reply:
These all got renamed to top level private instance variable on the class.